### PR TITLE
change loss parameter callback

### PR DESCRIPTION
**Pull request by yHarai**
# 概要
- lossのパラメータ, weightをcallbackで変更できるようにする
# 主な変更箇所
- keras_callbacks.py
  - lossのparameterを変更するcallbackクラスを追加
  - loss_weightsを変更するcallbackクラスを追加 (2021/12/27)
- losses.py
  - lossのパラメータをTensor（tf.Variable）に変更
  - 理由：callback内でパラメータを書き換える際、np.arrayの状態で書き換えても(setattr)、lossの計算時に書き換えたパラメータが反映されていなかった。keras.backend.set_valueで書き換えることでlossの計算時に書き換えたパラメータを反映することができたが、set_valueで値を書き換えるためにはパラメータをTensorで保持する必要がある。そのため、lossのパラメータをtf.Variableで持つことにした。
- 複数lossを指定する場合にsegmentation_modelsのSumOfLossesではなくCompoundLossを使用するように変更 (2021/12/27)
  - loss_weightを指定できるようにするため。また、callbackで変更できるようにするため。
- example/yamls/segmentation.yaml
  - callbackでlossのパラメータを変更する際の例を追記
# 動作確認
- farmer/exampleのdetection以外がエラーなく動くことを確認。
- 今回追加したcallbackの片方/両方がyaml指定で動くことを確認。
- LossWeightScheduler
  - CompoundLossのcall時に変更後のweightが適用されていることをログで確認。
- LossParamScheduler
  - example/segmentation.yamlでDiceLossにcallbackを適用し、call時に変更後のbetaが適用されていることをログで確認。

### CODE REVIEW

File: farmer/domain/tasks/build_model_task.py [hunk 1]
```diff
@@ -216,19 +216,17 @@ def _do_compile_model_task(
...
+            loss_container = list()
+            
             if task_id == Task.CLASSIFICATION:
                 for i, loss_func in enumerate(loss_funcs.items()):
                     loss_name, params = loss_func
-                    if i == 0:
-                        if params is None:
-                            loss = getattr(keras.losses, loss_name)()
-                        else:
-                            loss = getattr(keras.losses, loss_name)(**params)
```
Comment (yHarai) [inline farmer/domain/tasks/build_model_task.py:226]: [説明]lossが単数・複数で分岐があったが、SumOfLossesを使わなくなったことにより分岐なしで共通化できるため削除。

File: farmer/ncc/callbacks/keras_callbacks.py [hunk 2]
```diff
@@ -5,9 +5,124 @@
...
 from ..utils import PostClient, MatPlotManager
 from ..metrics import iou_dice_val, generate_segmentation_result
-
+from ..losses import CompoundLoss
+from ..callbacks import functional as F
 from tensorflow import keras
+import tensorflow as tf
+
+class LossWeightsScheduler(keras.callbacks.Callback):
+    def __init__(self, model, schedulers):
```
Comment (yHarai) [inline farmer/ncc/callbacks/keras_callbacks.py:14]: [説明]init時にmodelを渡すのは最初にcallback適用可能かを判断したいため。（on_epoch_endでエラーを出すと気付くのに時間がかかってしまう。）

File: farmer/domain/tasks/build_model_task.py [hunk 3]
```diff
@@ -246,22 +244,26 @@ def _do_compile_model_task(
...
                         class_indexes=list(range(1, self.config.nb_classes)))
                     ],
 
-            model.compile(optimizer, loss, metrics)
+            # compound loss
+            if len(loss_container) > 1:
+                loss_weights = loss.get("loss_weights", dict())
+                w_l1 = loss_weights.get('l1', 1.)
+                w_l2 = loss_weights.get('l2', 1.)
+                loss_instance = losses.CompoundLoss(loss_container[0], loss_container[1], w_l1, w_l2)
```
Comment (aiorhiroki) [inline farmer/domain/tasks/build_model_task.py:265]: このlossの個数を可変長にしたくなるけど、分かりやすさ的にはこのままにしますか。
lossが3つ以上でl3があるときの実装は、実際に必要になってからで良いですよね。

**Review by aiorhiroki — APPROVED** [review]
素晴らしいコーディングですね！！

File: farmer/domain/tasks/build_model_task.py (hunk 3 above)
Comment (yHarai) [reply farmer/domain/tasks/build_model_task.py:265]: ご指摘の通り、ここは可変長にするか迷いましたが、分かりやすさを取りました。
今回はこのままでお願いいたします！

File: farmer/ncc/callbacks/keras_callbacks.py [hunk 4]
```diff
@@ -5,9 +5,124 @@
...
+        super(LossWeightsScheduler, self).__init__()
+        self.scheduler = list()
+        
+        for loss_index, scheduler in schedulers.items():
+            if hasattr(model.loss, f'w_{loss_index}'):
+                update_fn = getattr(F, scheduler['schedule_fn'])
+                self.scheduler.append((f'w_{loss_index}', update_fn, scheduler['params']))
+            else:
+                print(f'{model.loss.__name__} does not have attribute: w_{loss_index}. Loss must be <losses.CompoundLoss>')
+                raise AttributeError
```
Comment (yumion) [inline farmer/ncc/callbacks/keras_callbacks.py]: 好みの問題かもしれませんが、エラー文はprintではなくraiseの方で一緒に出したいです

```
raise AttributeError(f'{model.loss.__name__} does not have attribute: w_{loss_index}. Loss must be <losses.CompoundLoss>')
```

File: farmer/ncc/callbacks/keras_callbacks.py [hunk 5]
```diff
@@ -5,9 +5,124 @@
...
+        self.loss = self.loss_instance(model)
 
+        # set self.scheduler
+        for param_name, scheduler in schedule_target.items():
+            if hasattr(self.loss, param_name):
+                update_fn = getattr(F, scheduler['schedule_fn'])
+                self.scheduler.append((param_name, update_fn, scheduler['params']))
+            else:
+                print(f'{self.loss.__name__} does not have attribute: {param_name}')
+                raise AttributeError
```
Comment (yumion) [inline farmer/ncc/callbacks/keras_callbacks.py]: ここも

File: farmer/ncc/callbacks/keras_callbacks.py [hunk 6]
```diff
@@ -5,9 +5,124 @@
...
+    def loss_instance(self, model):
+        # compound loss
+        if isinstance(model.loss, CompoundLoss):
+            if model.loss.l1.__name__ == self.loss_name:
+                cls = model.loss.l1
+            elif model.loss.l2.__name__ == self.loss_name:
+                cls = model.loss.l2
+            else:
+                print(f'not compiled loss: {self.loss_name}')
+                raise NotImplementedError
```
Comment (yumion) [inline farmer/ncc/callbacks/keras_callbacks.py]: ここもですね
```
raise NotImplementedError(f'not compiled loss: {self.loss_name}')
```

File: farmer/ncc/callbacks/keras_callbacks.py [hunk 7]
```diff
@@ -5,9 +5,124 @@
...
+        # single loss
+        else:
+            if model.loss.__name__ == self.loss_name:
+                cls = model.loss
+            else:
+                print(f'not compiled loss: {self.loss_name}')
+                raise NotImplementedError
```
Comment (yumion) [inline farmer/ncc/callbacks/keras_callbacks.py]: ここも

**Review by yumion — COMMENTED** [review]
raiseのエラー文についてコメントしました

**Comment by yumion** [comment]
他はLGTMです！

File: farmer/ncc/callbacks/keras_callbacks.py (hunk 4 above)
Comment (yHarai) [reply farmer/ncc/callbacks/keras_callbacks.py]: 修正しました。
https://github.com/aiorhiroki/farmer.tf2/pull/9/commits/6426181e325caf4180b50c11b825eb7fec0c84e4

**Comment by Yuto-web** [comment]
僕からは特に指摘はありません。LGTMです。
